TripalCultivate/TripalCultivate-Phenotypes
# G4.47 datafile validator

**Issue #47 Data File Validator** 

## Motivation



This PR adds another validation scope or level. The Data File validator (could not simplify it it to just File since File is a Drupal reserved keyword/word) validates files provided to the file field in the importer.

This PR is dependent on (https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/42) Genus Validator

## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Updates the scopes array in the validateForm() method of the Importer to include the key/scope 'FILE'.
2. Adds a plugin into the plugin directory that corresponds to the FILE validator scope.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

This PR has automated test included to test the following:
1. Kernel Test: Test files for:
  - File is of the expected file type/format of tsv (default file type) or txt (alternative). 
  - A file with 0 file size.
  - A valid extension of tsv file format but is with incorrect file format/mime.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Install/Enable Tripal Cultivate Phenotypes Module.
2. Create test data
Create a project
Create a genus
Pair the project - genus in project prop using the genus term in the configuration as the type, alternatively with the module properly configured use the service below.

```
\Drupal::service('trpcultivate_phenotypes.genus_project')
  ->setGenusToProject(PROJECT ID, 'GENUS', TRUE);
```
3. Execute Tripal Job created by the install.
4. Configure module.
5. Load Tripal Cultivate PhenoShare Importer page.
mysite/admin/tripal/loaders/trpcultivate-phenotypes-share

6. Download a template file provided in the importer and use it as test file. Alternatively, sample test files are in this attachment.

[Test Files.zip](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/files/13444248/Test.Files.zip)
